File: 1D Dynamic Programming/377. Combination Sum IV/Solution.java
class Solution {
    private int dfs(int total, int[] nums, int[] dp){
        if (total == 0){
            return 1;
        }
        if (total < 0){
            return 0;
        }
        if (dp[total] != -1){
            return dp[total];
        }

        int count = 0;
        for (int num: nums){
            if (total < num) continue;
            count += dfs(total - num, nums, dp);
        }
        dp[total] = count;
        return dp[total];
    }
    public int combinationSum4(int[] nums, int target) {
        // int[] dp = new int[target + 1];
        // Arrays.fill(dp, -1);
        // return dfs(target, nums, dp);

        int[] dp = new int[target + 1];
        dp[0] = 1;
        
        for(int total=1; total<(target + 1); total++){
            int count = 0;
            for (int num: nums){
                if (total < num) continue;
                count += dp[total - num];
            }
            dp[total] = count;
        }
        return dp[target];
    }
}

File: 1D Dynamic Programming/377. Combination Sum IV/Solution.py
from typing import List

class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        # n = len(nums)
        # dp = [-1] * (target + 1)
        # def dfs(total):
        #     if total == 0:
        #         return 1
        #     if total < 0:
        #         return 0
        #     if dp[total] != -1:
        #         return dp[total]
            
        #     res = 0
        #     for num in nums:
        #         if total - num < 0:
        #             continue
        #         res += dfs(total - num)
        #     dp[total] = res
        #     return dp[total]
        # return dfs(target)

        dp = [0] * (target + 1)
        dp[0] = 1
        for total in range(1, target + 1):
            res = 0
            for num in nums:
                if total < num:
                    continue
                res += dp[total - num]
            dp[total] = res
        return dp[target]